dds_web/api/user.py

```diff
@@ -86,6 +86,16 @@ def post(self):
         except sqlalchemy.exc.OperationalError as err:
             raise ddserr.DatabaseError(message=str(err), alt_message="Unexpected database error.")
 
+        # check if the user invite should have expired and be deleted
+        if unanswered_invite:
+            # If actual time minus 168 hours (7 days) is greater than the time in the invite -> is expired, delete it
+            if (
+                datetime.datetime.utcnow() - datetime.timedelta(hours=168)
+            ) > unanswered_invite.created_at:
+                db.session.delete(unanswered_invite)
+                db.session.commit()
+                unanswered_invite = None
+
```

Comment:
This endpoint is where the invite is created, so we need to move this check to `dds_web/web/user.py`, the `confirm_invites` endpoint. That's where the user ends up when they click on an invite link.

Reply:
If you want we can look at it tomorrow after the daily and I can explain how I'm imagining the fix looking?

Reply:
Looking at this it seems that my suggestion for deleting the link is also quite complicated, so I think it's a good idea if we discuss this more after the daily and come up with a game plan and possibly do some pair coding or decide on an approach.

Reply:
But wouldn't that produce the same end result?

1. Researcher click on expired token and cannot register
2. Unit admin sent again invite
3. The function checks that their user had an old date and remove the row and generates new token
4. Researcher get email with a different token

The deletion either occurs between 1-2 or 2-3 But the unit won't get the error and the researcher gets the new email.
That said, it don't hurt to do the check also there!

Reply:
So where you put the code now (in `dds_web/api/user.py::AddUser`, not `dds_web/web/user.py::confirm_invite`) is only used when a user tries to invite another user, and when the invite is sent out as an email. So this code would only be executed if a user was already invited, and then they try to invite the user again before the user has accepted the invite. It wouldn't be executed when a user clicks on their invite link and their link has expired.

Reply:
Aha ok, I see what you mean now. Yes, this is another option, I'll have a look again tomorrow with a rested brain.

Reply:
Ok so I've had a look now and I get it now. I like it!

I suggest a few changes though, one is this:
- Move this code section to `dds_web/api/schemas/user_schemas.py::UnansweredInvite::return_invite`, after line 73 (the query for the invite).
- Change `datetime.timedelta(hours=168)` to `datetime.timedelta(hours=flask.current_app.config["INVITATION_EXPIRES_IN_HOURS"])`

---

* I also checked so that any ProjectInviteKeys are removed (they should ofc be, but just as a check, should also be included in tests), and they are, and associated with a new invite when it's created.
* And the following commands work exactly as they should:
```bash
> dds user add tester@mail.com --role Researcher
.....
ERROR    ⚠️ Failed to add user: This user was already added to the system. Specify the project you wish to give access to. ⚠️   


> dds user add tester@mail.com --role Researcher --project someunit00001
.....
ERROR    Failed to add user: User tester is already associated with the Project someunit00001 in this capacity.               


> dds project access fix tester@mail.com
.....
Project access updated for user 'tester@mail.com'.
```